### DISCUSSION TITLE
Select options based on repeater items

### DISCUSSION THREAD

**Discussion by joshembling** [filamentphp/filament#9370] [General]
I have a repeater with a shape like this: 

```php
Repeater::make('members')
    ->schema([
        TextInput::make('name')->required(),

        Repeater::make('list')
            ->schema([
                self::getLinksFieldset('issues'),
            ])
    ])
```

Within my top level repeater I have a `TextInput` 'name' field and within this repeater, a child level repeater with a select option. These options are based on the parent repeater items. 

So let's say I have 5 repeaters with names of 'Name1', 'Name2' etc. Then my select options will appear as:

```
1 => Name1
2 => Name2
```

I have split the select options into the following method:
```php
   public static function getLinksFieldset(string $block): Fieldset
    {
        return Fieldset::make('Links')
            ->schema([
                Select::make('linkId')
                    ->options(function (Get $get, $component) use ($block) {
                        $guides = $get('../../../../'.$block); // the top level repeater
                        $key = explode('.', $component->getStatePath())[5]; // the repeater key

                        $options = [];

                        foreach ($guides as $idx => $guide) {
                            if ($idx !== $key && ! is_null($guide['id']) && ! is_null($guide['name'])) {
                                $id = (int) $guide['id'];
                                $options[$id] = $guide['name'];
                            }
                        }

                        return $options;
                    })
                    ->live()
                    ->helperText('Optionally, select the piece of information you would like to link to.'),
                TextInput::make('linkText')
                    ->visible(fn (Get $get): ?bool => $get('link')),
            ]);
    }
```

An id is generated for each repeater item like so: 
```php 
    public static function generateNextId(?array $repeater_items)
    {
        $highestId = 0;

        if (is_null($repeater_items)) {
            return 0;
        }

        foreach ($repeater_items as $innerArray) {
            $highestId = max($highestId, $innerArray['id']);
        }

        return $highestId + 1;
    }
```

All of this works apart from when an item is deleted, or a new item is added. 

When an item is deleted, it gets removed from the select options, but all of the other options selected within the child repeater are shifted up one level e.g. 

```php
// before
1 => Name1
2 => Name2
3 => Name3
4 => Name4

// on deletion of item 2
1 => Name1
empty => empty // correct
4 => Name3 // incorrect
blank => Name4 // incorrect
```

This is corrected on save. So the problem is watching for the live changes of the repeater item. 

This removes the item: 
```php
    public static function deleteRelatedLinksWhenRepeaterItemIsDeleted(Action $action, string $list_key): Action
    {
        return $action
            ->requiresConfirmation()
            ->modalDescription('Are you sure you\'d like to delete this item? This cannot be undone.')
            ->before(function ($component, $arguments, $state) use ($list_key) {
                $item = $state[$arguments['item']];

                foreach ($state as $repeater_idx => $repeater_item) {
                    if (array_key_exists($list_key, $repeater_item)) {
                        foreach ($repeater_item[$list_key] as $list_idx => $list_items) {

                            if (array_key_exists('linkId', $list_items)) {

                                if ($list_items['linkId'] == $item['id']) {
                                    $state[$repeater_idx][$list_key][$list_idx]['linkId'] = null;
                                    $state[$repeater_idx][$list_key][$list_idx]['linkText'] = null;
                                }
                            }
                        }
                    }
                }

                $component->state($state);
            });
    }
```

But the problem remains where the index isn't corrected until I save. 

TLDR: How do I watch for global changes on any repeater item to update select options that rely on said top level repeater items?
